src/primitives/cs_helpers.rs

```diff
@@ -25,46 +29,93 @@ pub fn constraint_evaluation(
     challenge_power_offset: usize,
     quotient: &mut [F],
     domain_size: usize,
+    is_specialized: bool,
 ) -> CudaResult<()> {
     assert_eq!(quotient.len(), 2 * domain_size);
-    assert!(gates.is_empty() == false);
-
-    let variable_columns_matrix = DeviceMatrix::new(
-        unsafe { DeviceSlice::from_slice(variable_columns.as_ref()) },
-        domain_size,
-    );
-    let witness_columns_matrix = DeviceMatrix::new(
-        unsafe { DeviceSlice::from_slice(witness_columns.as_ref()) },
-        domain_size,
-    );
-    let constant_columns_matrix = DeviceMatrix::new(
-        unsafe { DeviceSlice::from_slice(constant_columns.as_ref()) },
-        domain_size,
-    );
+    assert!(!gates.is_empty());
 
     let mut d_challenge = svec!(2);
     mem::d2d(&challenge.c0.inner[..], &mut d_challenge[..1])?;
     mem::d2d(&challenge.c1.inner[..], &mut d_challenge[1..])?;
     let challenge = unsafe { DeviceSlice::from_slice(&d_challenge[..]) };
     let challenge = unsafe { challenge.transmute::<VectorizedExtensionField>() };
 
-    let quotient = unsafe { DeviceSlice::from_mut_slice(quotient.as_mut()) };
-    let mut quotient_matrix = DeviceMatrixMut::new(
-        unsafe { quotient.transmute_mut::<VectorizedExtensionField>() },
-        domain_size,
-    );
-
-    if_not_dry_run! {
-        boojum_cuda::gates::evaluate_gates(
-            &gates,
-            &variable_columns_matrix,
-            &witness_columns_matrix,
-            &constant_columns_matrix,
-            challenge,
-            &mut quotient_matrix,
-            challenge_power_offset as u32,
-            get_stream(),
-        ).map(|_| ())
+    let variables_slice = unsafe { DeviceSlice::from_slice(variable_columns.as_ref()) };
+    let witnesses_slice = unsafe { DeviceSlice::from_slice(witness_columns.as_ref()) };
+    let constants_slice = unsafe { DeviceSlice::from_slice(constant_columns.as_ref()) };
+    let quotient_slice = unsafe {
+        DeviceSlice::from_mut_slice(quotient.as_mut()).transmute_mut::<VectorizedExtensionField>()
+    };
+    const STREAMS_COUNT: usize = 4;
+    assert!(STREAMS_COUNT <= NUM_AUX_STREAMS_AND_EVENTS);
+    const BLOCK_SIZE: usize = 128;
+    let l2_size = _l2_cache_size();
+    let capability = _compute_capability_major();
+    let cols_count =
+        (variables_slice.len() + witnesses_slice.len() + constants_slice.len()) / domain_size + 2;
+    let chunk_rows =
+        l2_size / (STREAMS_COUNT * size_of::<F>() * cols_count) / BLOCK_SIZE * BLOCK_SIZE;
```

Comment:
i had to scratch my head for several minutes to reassure myself the order/associativity gives what you want, but i think it does :P